Feat: Email Validation

m2rads opened the pull request:
This PR addresses issue #104 


## Changes
- Add Mailosaur for email generation, email rendering, 2fa and content validation. 
- Add a new tool called `sleep_miliseconds` that helps AI to wait for a condition to be met before proceeding

m2rads commented:
@slavingia FYI I ended up using Mailosaur as it provides more features and is faster.

m2rads commented on packages/shortest/package.json:
```diff
@@ -29,36 +29,38 @@
     "build:js": "esbuild src/index.ts --bundle --platform=node --format=esm --outfile=dist/index.js --external:esbuild --external:punycode --external:playwright --external:@anthropic-ai/sdk --external:expect --external:dotenv",
     "build:cjs": "esbuild src/index.ts --bundle --platform=node --format=cjs --outfile=dist/index.cjs --external:esbuild --external:punycode --external:playwright --external:@anthropic-ai/sdk --external:expect --external:dotenv",
-    "build:cli": "esbuild src/cli/bin.ts src/cli/setup.ts --bundle --platform=node --format=esm --outdir=dist/cli --metafile=dist/meta-cli.json --external:fsevents --external:chokidar --external:glob --external:esbuild --external:events --external:path --external:fs --external:util --external:stream --external:os --external:assert --external:url --external:playwright --external:@anthropic-ai/sdk  ...
+    "build:cli": "esbuild src/cli/bin.ts src/cli/setup.ts --bundle --platform=node --format=esm --outdir=dist/cli --metafile=dist/meta-cli.json --external:fsevents --external:chokidar --external:glob --external:esbuild --external:events --external:path --external:fs --external:util --external:stream --external:os --external:assert --external:url --external:playwright --external:@anthropic-ai/sdk  ...
```
Here we are just marking some more nodejs built ins as external so that they will work with Mailosaur implementation:. Here is the list: 

https
http
net
tls
crypto

m2rads commented on packages/shortest/src/ai/prompts/index.ts:
```diff
@@ -42,6 +43,11 @@ IMPORTANT GLOBAL RULES:
    - All expectations listed in the test instructions must be fulfilled.
    - If any expectation is not met, the test case must be marked as failed.
 
+8. **Testing Email**:
```
Add a new rule to system prompt to instruct AI how to use Email validation

m2rads commented on packages/shortest/src/ai/tools/index.ts:
```diff
@@ -29,6 +29,42 @@ export const AITools = [
       required: ["action", "username", "password"]
     }
   },
+  {
+    name: "render_email",
+    description: "Fetch, decode and render received email in browser",
+    input_schema: {
+      type: "object",
+      properties: {
+        action: {
+          type: "string",
+          enum: ["render_email"],
+          description: "Render the received email in a new tab"
+        }
+      },
+      required: ["action"]
+    }
+  },
+  {
+    name: "sleep_milliseconds",
+    description: "Pause test execution for specified duration",
+    input_schema: {
+      type: "object",
+      properties: {
+        action: {
+          type: "string",
+          enum: ["sleep_milliseconds"],
+          description: "The action to perform"
+        },
+        duration: {
+          type: "number",
+          description: "Duration to sleep in milliseconds (e.g. 5000 for 5 seconds)",
+          minimum: 0,
+          maximum: 60000
+        }
+      },
+      required: ["action", "duration"]
+    }
+  },
```
Added two new tools here: render_email and sleep_miliseconds 

render_email is used to execute Mailosaur email validation in `browser-tools.ts` file. 
`sleep_miliseconds` is used to help AI to wait for a condition to be met. Either instructed by the user or based on AI judgement.

Why both of them needed in this PR: 
getting the latest email requires AI to wait for a few seconds to so that the user receives the email, otherwise the test case might fail prematurely

m2rads commented on packages/shortest/src/browser/core/browser-tool.ts:
```diff
@@ -389,6 +395,69 @@ export class BrowserTool extends BaseBrowserTool {
           }
         }
 
+        case 'sleep_milliseconds': {
+          const defaultDuration = 1000;
+          const maxDuration = 60000;
+          let duration = input.duration ?? defaultDuration;
+          
+          // Enforce maximum duration
+          if (duration > maxDuration) {
+            console.warn(`Requested sleep duration ${duration}ms exceeds maximum of ${maxDuration}ms. Using maximum.`);
+            duration = maxDuration;
+          }
+
+          // Convert to seconds for logging
+          const seconds = Math.round(duration / 1000);
+          console.log(`⏳ Waiting for ${seconds} second${seconds !== 1 ? 's' : ''}...`);
+          
+          await this.page.waitForTimeout(duration);
+          output = `Finished waiting for ${seconds} second${seconds !== 1 ? 's' : ''}`;
+          break;
+        }
```
Implementation for sleep_miliseconds: max time is 60 seconds and min is 1 second

m2rads commented on packages/shortest/src/browser/core/browser-tool.ts:
```diff
@@ -389,6 +395,69 @@ export class BrowserTool extends BaseBrowserTool {
+              throw new ToolError('Mailosaur configuration required');
+            }
+            this.mailosaurTool = new MailosaurTool(this.config.mailosaur);
+          }
+
+          const newPage = await this.page.context().newPage();
+
+          try {
+            const email = await this.mailosaurTool.getLatestEmail();
+
+            // Render email in new tab
+            await newPage.setContent(email.html, {
+              waitUntil: 'domcontentloaded'
+            });
+
+            await newPage.waitForLoadState('load', {
+              timeout: 5000
+            }).catch(e => {
+              console.log('⚠️ Load timeout, continuing anyway');
+            });
+
+            // Switch focus
+            this.page = newPage;
+
+            output = `Email received successfully. Navigated to new tab to display email: ${email.subject}`;
+            metadata = {
+              window_info: {
+                title: email.subject,
+                content: email.html,
+                size: this.page.viewportSize() || { width: this.width, height: this.height }
+              }
+            };
+
+            break;
+          } catch (error) {
+            await newPage.close();
+            throw new ToolError(`Failed to render email: ${error}`);
+          }
+        }
+
```
Implementation for render_email: This tool will grab the latest email from Mailosaur in HTML format and will render it in a new tab. It will then give the control back to AI to validate the email. 

Why render it in a new tab? 
Some users may want to test the links in the email. This gives AI a chance to interact with the email and click on it since it's in the browser. 

Just a note that this is indeterministic. Perhaps we could implement a mechanism to manually compare the content of the email and the received email in the inbox.

m2rads commented on packages/shortest/src/browser/integrations/mailosaur.ts:
This file is the implementation of Mailosaur in terms of grabbing the latest email from inbox. We can use this file to add more features like SMS validation and other types of email validations.

m2rads commented on packages/shortest/src/types/config.ts:
```diff
@@ -6,4 +6,9 @@ export interface ShortestConfig {
   baseUrl: string;
   testDir: string | string[];
   anthropicKey: string;
+  mailosaur?: {
+    apiKey: string;
+    serverId: string;
+    emailAddress: string;
+  };
```
apiKey and serverId are necessary but not emailAddress. I will remove it.

m2rads reviewed (commented):
Hi @slavingia . Here's my implementation for Email Validation. Below is a simple test case in Helper. I added explanation for important parts of the code for you. Let me know if you'd like me explain any parts in more details. Before I finalize my PR please let me know if you'd like to adjust something. Here's the demo: 


https://github.com/user-attachments/assets/3f7b6096-b77d-4cd6-8c56-c67af73768c2

m2rads commented on packages/shortest/src/browser/core/browser-tool.ts:
```diff
@@ -389,6 +395,100 @@ export class BrowserTool extends BaseBrowserTool {
+
+            break;
+          } catch (error: unknown) {
+            await newPage.close();            
+            const errorMessage = error instanceof Error ? error.message : String(error);
+
+            if (errorMessage.includes('Email content missing')) {
+              return {
+                output: `Email was found but content is missing. This might be due to malformed email. Moving to next test.`,
+                error: 'EMAIL_CONTENT_MISSING'
+              };
+            }
+
+            if (errorMessage.includes('Mailosaur email address is required')) {
+              return {
+                output: `Email address is required but was not provided.`,
+                error: 'EMAIL_ADDRESS_MISSING'
+              };
+            }
+
+            if (errorMessage.includes('No matching messages found')) {
+              return {
+                output: `No email found for ${input.email}. The email might not have been sent yet or is older than 1 hour. Moving to next test.`,
+                error: 'EMAIL_NOT_FOUND'
+              };
+            }
+
+            // Generic error case
+            return {
+              output: `Failed to fetch or render email: ${errorMessage}. Moving to next test.`,
+              error: 'EMAIL_OPERATION_FAILED'
+            };
+          }
+        }
+
```
A bit more robust error handling to handle validation failures more accurately.

m2rads commented on packages/shortest/CHANGELOG.md:
```diff
@@ -5,6 +5,15 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.1.2] - 2024-12-26
+
+### Added  
+- Added support for Mailosaur email validation
+- Added email rendering feature in the browser
+- Added sleep_milliseconds tool to add delays in the test execution
+- Added more robust error handling for Mailosaur email validation
```
Adding the features that was added to in this PR. Since we have added a new feature, I used the wording `Added`. Maybe we can have a consistent wording for this file!?

m2rads commented on packages/shortest/src/ai/prompts/index.ts:
```diff
@@ -42,6 +43,14 @@ IMPORTANT GLOBAL RULES:
    - All expectations listed in the test instructions must be fulfilled.
    - If any expectation is not met, the test case must be marked as failed.
 
+8. **Testing Email**:
+   - If you need to test a condition that involves seeing the contents of an email, use the "render_email" tool.
+   - This tool will grab the latest email from the email address given to you and will render it in a new tab for you to see.
+   - Once you are done with validating the email, navigate back to the original tab.
+   - You MUST pass the email address that is given to you to the tool as a parameter otherwise it will fail.
+   - If no email address is given to you for this test, you should fail the test.
+   - For email validation, you MUST always use 'Click' and 'Mouse' action instead of using keyboard shortcuts.
+
```
Add a new rule to system prompt to instruct AI how to use Email validation. Specifically instructing AI to pass email address as ActionInput and use 'Click' and 'Mouse' action instead of using keyboard shortcuts because shortcuts may cause unexpected events when dealing with an Inbox UI.

m2rads reviewed (commented):
Ready for review!

slavingia commented on app/__tests__/dashboard.test.ts:
```diff
@@ -3,19 +3,14 @@ import { clerk, clerkSetup } from "@clerk/testing/playwright";
 
 let frontendUrl = process.env.SHORTEST_TEST_BASE_URL ?? "http://localhost:3000";
 
-shortest.beforeAll(async ({ page }) => {
```
Self-review would be helpful to me. Why isn't this needed?

slavingia commented on packages/shortest/src/ai/tools/index.ts:
```diff
@@ -29,6 +29,42 @@ export const AITools = [
     }
   },
+  {
+    name: "render_email",
```
```suggestion
    name: "check_email",
```

?

slavingia commented on packages/shortest/src/ai/tools/index.ts:
```diff
@@ -29,6 +29,42 @@ export const AITools = [
+    }
+  },
+  {
+    name: "sleep_milliseconds",
```
```suggestion
    name: "sleep",
```

Cleaner?

slavingia commented on app/__tests__/dashboard.test.ts:
```diff
@@ -3,19 +3,14 @@ import { clerk, clerkSetup } from "@clerk/testing/playwright";
+
+shortest("Verify that users can generate test cases")
+.expect("It takes up to 20 seconds for the tests to be generated")
+.expect("You should validate that the vitest tests are generated and shown in the UI")
```
Would be great to add a (feature and) test case for email checking here.

m2rads commented:
@slavingia I addressed your reviews in this PR but it got a bit crowded with the previous self-review. I will close this and open a new one to start a clean self-review for Razvan.

slavingia commented:
IMO better to resolve comments and have the same PR so a reviewer can see if their comments were addressed.
